### PR TITLE
compiler: support neo-debugger

### PR DESCRIPTION
**Pull request by fyrchik**
Support debug information export.
There is no room in specification to support multiple returns, so `Any` (`Array`) type is used.

There are still things to do. Those can be good first issues. E.g. add more sequence points, i.e. places where debugger can stop (binary operators, switch cases, ...).

### CODE REVIEW

**Comment by roman-khimov** [comment]
>  There is no room in specification to support multiple returns, so Any (Array) type is used.

I think it's perfect time then to submit an issue for the spec!

File: pkg/compiler/debug.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+// EventDebugInfo represents smart-contract's event debug information.
+type EventDebugInfo struct {
+	ID string `json:"id"`
+	// Name is a human-readable event name in a format "{namespace}-{name}".
+	Name       string       `json:"name"`
+	Parameters []debugParam `json:"parameters"`
+}
+
+type debugRange struct {
```
Comment (roman-khimov) [inline pkg/compiler/debug.go:54]: Why is it unexported? `MethodDebugInfo` is exported and has exported field that uses this type.

File: pkg/compiler/debug.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	Start uint16
+	End   uint16
+}
+
+type debugParam struct {
```
Comment (roman-khimov) [inline pkg/compiler/debug.go:59]: Same thing here.

File: pkg/compiler/debug.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	Documents  []string          `json:"documents"`
+	Methods    []MethodDebugInfo `json:"methods"`
+	Events     []EventDebugInfo  `json:"events"`
+}
+
+// MethodDebugInfo represents smart-contract's method debug information.
+type MethodDebugInfo struct {
+	ID string `json:"id"`
+	// Name is the name of the method in a format "{namespace},{display-name}"
+	Name string `json:"name"`
```
Comment (roman-khimov) [inline pkg/compiler/debug.go:25]: If it has two components to it, why a `string` then?

File: pkg/compiler/debug.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	// Range is the range of smart-contract's opcodes corresponding to the method.
+	Range debugRange `json:"range"`
+	// Parameters is a list of method's parameters.
+	Parameters []debugParam `json:"params"`
+	// ReturnType is method's return type.
+	ReturnType string   `json:"return-type"`
+	Variables  []string `json:"variables"`
+	// SeqPoints is a map between source lines and byte-code instruction offsets.
+	// It has following format: "{address}[{document-index}]{start-line}:{start-column}-{end-line}:{end-column}".
+	SeqPoints []string `json:"sequence-points"`
```
Comment (roman-khimov) [inline pkg/compiler/debug.go:35]: This is a very complex thing packed into a `string`, why? It's OK for JSON to have it packed like that, but I'd rather have some more convenient representation here.

File: pkg/compiler/debug.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+func (c *codegen) scReturnTypeFromScope(scope *funcScope) string {
+	results := scope.decl.Type.Results
+	switch results.NumFields() {
+	case 0:
+		return "Void"
+	case 1:
+		return c.scTypeFromExpr(results.List[0].Type)
+	default:
+		// multiple return values are not supported in debugger
```
Comment (roman-khimov) [inline pkg/compiler/debug.go:135]: Have you submitted an issue for it?

File: pkg/compiler/codegen.go [hunk 6]
```diff
@@ -1291,9 +1281,29 @@ func CodeGen(info *buildInfo) ([]byte, error) {
...
+
+// CodeGen compiles the program to bytecode.
+func CodeGen(info *buildInfo) ([]byte, error) {
+	pkg := info.program.Package(info.initialPackage)
+	c := newCodegen(info, pkg)
+
+	if err := c.compile(info, pkg); err != nil {
+		return nil, err
 	}
+
```
Comment (roman-khimov) [inline pkg/compiler/codegen.go:1326]: Useless line

File: cli/smartcontract/smart_contract.go [hunk 7]
```diff
@@ -92,6 +92,10 @@ func NewCommands() []cli.Command {
 						Name:  "debug, d",
 						Usage: "Debug mode will print out additional information after a compiling",
 					},
+					cli.StringFlag{
+						Name:  "emitdebug",
```
Comment (roman-khimov) [inline cli/smartcontract/smart_contract.go:96]: Short form? And maybe it's better to rename `debug` flag into `verbose` or something like that and use `debug` here.

File: cli/smartcontract/smart_contract.go [hunk 8]
```diff
@@ -348,6 +352,8 @@ func contractCompile(ctx *cli.Context) error {
 	o := &compiler.Options{
 		Outfile: ctx.String("out"),
 		Debug:   ctx.Bool("debug"),
+
```
Comment (roman-khimov) [inline cli/smartcontract/smart_contract.go:355]: Blank line.

File: cli/smartcontract/smart_contract.go (hunk 7 above)
Comment (fyrchik) [reply cli/smartcontract/smart_contract.go:96]: https://github.com/nspcc-dev/neo-go/pull/829

File: pkg/compiler/debug.go (hunk 5 above)
Comment (fyrchik) [reply pkg/compiler/debug.go:142]: https://github.com/neo-project/neo-debugger/issues/17

**Comment by fyrchik** [comment]
> I guess to be compatible with `neo-debugger` we also need to zip the resulting JSON.

`neo-debugger` supports both JSON and ZIP formats.
We can make both `.zip`/`.json` depending on the provided filename. I don't want to drop JSON because it is easier to look at and debug.

**Comment by roman-khimov** [comment]
If if works with plain JSON too then we're probably fine, adding zip compression into our code doesn't make me happy either.

**Pull request merged**